Mathlib/Algebra/Order/BigOperators/Expect.lean

```diff
@@ -0,0 +1,217 @@
+lemma expect_mul_sq_le_sq_mul_sq (s : Finset ι) (f g : ι → R) :
+    (𝔼 i ∈ s, f i * g i) ^ 2 ≤ (𝔼 i ∈ s, f i ^ 2) * 𝔼 i ∈ s, g i ^ 2 := by
+  simp only [expect, smul_pow, inv_pow, smul_mul_smul_comm, ← sq]
+  gcongr
+  exact sum_mul_sq_le_sq_mul_sq ..
+
+end LinearOrderedCommSemiring
+end Finset
+
+open Finset
+
+namespace Fintype
+variable [Fintype ι] [Fintype κ]
+
+section OrderedAddCommMonoid
+variable [OrderedAddCommMonoid α] [Module ℚ≥0 α] {f : ι → α}
+
+lemma expect_eq_zero_iff_of_nonneg [Nonempty ι] (hf : 0 ≤ f) : 𝔼 i, f i = 0 ↔ f = 0 := by
+  simp [expect, sum_eq_zero_iff_of_nonneg hf, univ_nonempty.ne_empty]
+
+lemma expect_eq_zero_iff_of_nonpos [Nonempty ι] (hf : f ≤ 0) : 𝔼 i, f i = 0 ↔ f = 0 := by
+  simp [expect, sum_eq_zero_iff_of_nonpos hf, univ_nonempty.ne_empty]
+
+end OrderedAddCommMonoid
+end Fintype
+
+open Finset
+
+namespace Mathlib.Meta.Positivity
+open Qq Lean Meta Finset
+open scoped BigOperators
+
+/-- Positivity extension for `Finset.expect`. -/
+@[positivity Finset.expect _ _]
+def evalFinsetExpect : PositivityExt where eval {u α} zα pα e := do
+  match e with
+  | ~q(@Finset.expect $ι _ $instα $instmod $s $f) =>
+    let (lhs, _, (rhs : Q($α))) ← lambdaMetaTelescope f
+    let so : Option Q(Finset.Nonempty $s) ← do
+      try
+        let _fi ← synthInstanceQ q(Fintype $ι)
+        let _no ← synthInstanceQ q(Nonempty $ι)
+        match s with
+        | ~q(@univ _ $fi) => pure (some q(Finset.univ_nonempty (α := $ι)))
+        | _ => pure none
+      catch _ => do
+        let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) | pure none
+        pure (some (.fvar fv))
+    match ← core zα pα rhs, so with
+    | .nonnegative pb, _ => do
+      let instαordmon ← synthInstanceQ q(OrderedAddCommMonoid $α)
+      let instαordsmul ← synthInstanceQ q(PosSMulMono ℚ≥0 $α)
+      assumeInstancesCommute
+      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars lhs pb
+      return .nonnegative q(@expect_nonneg $ι $α $instαordmon $instmod $s $f $instαordsmul
+        fun i _ ↦ $pr i)
+    | .positive pb, .some (fi : Q(Finset.Nonempty $s)) => do
+      let instαordmon ← synthInstanceQ q(OrderedCancelAddCommMonoid $α)
+      let instαordsmul ← synthInstanceQ q(PosSMulStrictMono ℚ≥0 $α)
+      assumeInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars lhs pb
+      return .positive q(@expect_pos $ι $α $instαordmon $instmod $s $f $instαordsmul
+        (fun i _ ↦ $pr i) $fi)
+    | _, _ => pure .none
+  | _ => throwError "not Finset.expect"
+
+example (n : ℕ) (a : ℕ → ℚ) : 0 ≤ 𝔼 j ∈ range n, a j^2 := by positivity
+example (a : ULift.{2} ℕ → ℚ) (s : Finset (ULift.{2} ℕ)) : 0 ≤ 𝔼 j ∈ s, a j^2 := by positivity
+example (n : ℕ) (a : ℕ → ℚ) : 0 ≤ 𝔼 j : Fin 8, 𝔼 i ∈ range n, (a j^2 + i ^ 2) := by positivity
+example (n : ℕ) (a : ℕ → ℚ) : 0 < 𝔼 j : Fin (n + 1), (a j^2 + 1) := by positivity
+example (a : ℕ → ℚ) : 0 < 𝔼 j ∈ ({1} : Finset ℕ), (a j^2 + 1) := by
+  have : Finset.Nonempty {1} := singleton_nonempty 1
+  positivity
```

Comment:
This is a bit weird to me - can't positivity figure out that {1} is nonempty?

Reply:
Well spotted. This was accidentally the pre-`proveFinsetNonempty` version of the extension